File: plugins/plugin-auto-trader/rust/src/actions/run_backtest.rs
use async_trait::async_trait;
use serde_json::{json, Value};

use crate::service::TradingService;
use crate::types::TradingStrategy;
use crate::{Action, ActionExample, ActionResult};

pub struct RunBacktestAction;

fn parse_strategy(val: &str) -> TradingStrategy {
    match val.to_lowercase().as_str() {
        "rule_based" | "rulebased" | "rule-based" | "technical" => TradingStrategy::RuleBased,
        "llm" | "llm_driven" | "ai" => TradingStrategy::LlmDriven,
        _ => TradingStrategy::Random,
    }
}

#[async_trait]
impl Action for RunBacktestAction {
    fn name(&self) -> &str {
        "RUN_BACKTEST"
    }

    fn similes(&self) -> Vec<&str> {
        vec!["BACKTEST", "SIMULATE_STRATEGY", "TEST_STRATEGY"]
    }

    fn description(&self) -> &str {
        "Run a backtest for a given strategy over a specified period."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        _message: &Value,
        state: &Value,
        service: Option<&TradingService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "TradingService is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let strategy = state
            .get("strategy")
            .and_then(|v| v.as_str())
            .map(parse_strategy)
            .unwrap_or(TradingStrategy::Random);

        let period = state
            .get("period_days")
            .and_then(|v| v.as_u64())
            .unwrap_or(30) as u32;

        let result = svc.run_backtest(strategy, period).await;

        let text = format!(
            "Backtest Results ({}, {} days):\n  Trades: {}\n  Total PnL: ${:.2}\n  Win Rate: {:.1}%\n  Max Drawdown: {:.2}%\n  Sharpe Ratio: {:.2}",
            strategy,
            period,
            result.trades.len(),
            result.total_pnl,
            result.win_rate * 100.0,
            result.max_drawdown * 100.0,
            result.sharpe_ratio,
        );

        ActionResult {
            success: true,
            text,
            data: Some(json!(result)),
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "Backtest the random strategy over 30 days".to_string(),
            agent_response: "Backtest Results (Random, 30 days): ...".to_string(),
        }]
    }
}

File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/actions/run_backtest.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_auto_trader.actions.common import ActionResult, Message
from elizaos_plugin_auto_trader.service import TradingService
from elizaos_plugin_auto_trader.types import TradingStrategy


def _parse_strategy(val: str) -> TradingStrategy:
    lower = val.lower()
    if lower in ("rule_based", "rulebased", "rule-based", "technical"):
        return TradingStrategy.RULE_BASED
    if lower in ("llm", "llm_driven", "ai"):
        return TradingStrategy.LLM_DRIVEN
    return TradingStrategy.RANDOM


@dataclass
class RunBacktestAction:
    @property
    def name(self) -> str:
        return "RUN_BACKTEST"

    async def validate(self, _message: Message, _state: dict) -> bool:  # type: ignore[type-arg]
        return True

    async def handler(
        self,
        _message: Message,
        state: dict,  # type: ignore[type-arg]
        service: TradingService | None = None,
    ) -> ActionResult:
        if service is None:
            return ActionResult(False, "TradingService is not available.", "missing_service")

        strategy = _parse_strategy(state.get("strategy", "random"))
        period = int(state.get("period_days", 30))

        result = await service.run_backtest(strategy, period)

        text = (
            f"Backtest Results ({strategy.value}, {period} days):\n"
            f"  Trades: {len(result.trades)}\n"
            f"  Total PnL: ${result.total_pnl:.2f}\n"
            f"  Win Rate: {result.win_rate * 100:.1f}%\n"
            f"  Max Drawdown: {result.max_drawdown * 100:.2f}%\n"
            f"  Sharpe Ratio: {result.sharpe_ratio:.2f}"
        )

        return ActionResult(True, text, data=result)
